### PR TITLE
Add Qthreads support

### PR DESCRIPTION
**Pull request by janciesko**
## Pull Request Description
Adds Qthreads support to MPICH. This allows hybrid MPI+Qthreads applications, similarly to the existing support of MPI+Argobots.

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/authorship

**Comment by janciesko** [comment]
Any TODO's on my end in regard to this PR?

**Review by hzhou — APPROVED** [review]
LGTM as long as tests are clean.

**Comment by hzhou** [comment]
test:mpich/ch3/tcp
test:mpich/ch4/ofi

**Comment by raffenet** [comment]
test:mpich/ch4/thread-package

**Comment by raffenet** [comment]
test:mpich/ch4/thread-package

File: src/mpl/include/mpl_thread_qthreads.h
Comment (raffenet) [inline src/mpl/include/mpl_thread_qthreads.h:19]: @janciesko is `qthread_spinlock_t` only available in Qthreads 1.19 and up? I had trouble getting that tarball to install on our Jenkins nodes. 1.18 installed fine, but is met with compilation errors.

Comment (janciesko) [reply src/mpl/include/mpl_thread_qthreads.h:19]: Yes, it requires version `1.19` or later. `1.19` added that type in order to support what MPICH required for Qthreads support.

Comment (janciesko) [reply src/mpl/include/mpl_thread_qthreads.h:19]: @raffenet Feel free to post the error that occurs when using 1.19 in the Jenkins job. It theory, it should work comparably to 1.18.

Comment (raffenet) [reply src/mpl/include/mpl_thread_qthreads.h:19]: It was actually errors when trying to build 1.19 from the release tarball. Here is the output.
```
System Characteristics:
       Target Style:
         Multi-node: no
       Topology API: libnumaV2
        Qtimer type: clock_gettime
     Aligned_t size: 8 (aligned on 8 byte boundaries)
 Default Stack size: 16kB

Safety/Debugging:
   Sanity assert()s: no
    Check alignment: no
          Profiling: none
   Debugging Output: no
        Guard Pages: no
          Precaching: no

Speed:
          Scheduler: sherwood
         Sinc Style: donecount
        Alloc Style: base
      Barrier Style: feb
   Dictionary Style: shavit
    Lazy Thread IDs: yes
       Pools/caches: memory
Increments/CAS/FEBs: Compiler Builtin (both), lock-based hash

Miscellany:
      Eureka Events: no

[raffenet@pmrs-centos64-240-04]/var/tmp/raffenet/qthreads-1.19% make -j8 install
CDPATH="${ZSH_VERSION+.}:" && cd . && /bin/sh '/var/tmp/raffenet/qthreads-1.19/config/missing' aclocal-1.16 -I config
CDPATH="${ZSH_VERSION+.}:" && cd . && /bin/sh '/var/tmp/raffenet/qthreads-1.19/config/missing' autoconf
 cd . && /bin/sh /var/tmp/raffenet/qthreads-1.19/config/missing automake-1.16 --foreign
/bin/sh ./config.status --recheck
running CONFIG_SHELL=/bin/sh /bin/sh ./configure --prefix=/nfs/gce/projects/pmrs/opt/qthreads-1.19 --no-create --no-recursion
./configure: line 4: .: filename argument required
.: usage: . filename [arguments]
make: *** [config.status] Error 2
make: *** Waiting for unfinished jobs....
[raffenet@pmrs-centos64-240-04]/var/tmp/raffenet/qthreads-1.19%
```
Running `autogen.sh` seems to have cleared up the issue locally, so I'll install that and give it a shot.

**Comment by raffenet** [comment]
test:mpich/ch4/thread-package

**Comment by raffenet** [comment]
@janciesko can you update `test/mpi/include/mpithreadtest.h` with support for Qthreads? The testsuite needs to be configured/built separately. You can do so like this:
```
cd test/mpi
./configure --with-mpi=<path/to/mpich> --with-thread-package=qthreads --with-qthreads=<path/to/qthreads>
```
Then to test just choose an arbitrary directory and see if the tests build:
```
cd attr
make testing
```

File: src/mpl/include/mpl_thread_qthreads.h
Comment (janciesko) [reply src/mpl/include/mpl_thread_qthreads.h:19]: Reproduced. Working on solution.

Comment (janciesko) [reply src/mpl/include/mpl_thread_qthreads.h:19]: Resolved. @raffenet, the tarball has been updated.

**Comment by raffenet** [comment]
test:mpich/ch4/thread-package

**Review by raffenet — APPROVED** [review]
LGTM. Tests are now running. Will let the current review job complete before merge. We can address any lingering failures in follow-up PRs.

**Comment by raffenet** [comment]
test:mpich/warnings

**Comment by janciesko** [comment]
Ack, thanks!

**Comment by raffenet** [comment]
test:mpich/ch4/thread-package

**Comment by raffenet** [comment]
test:mpich/ch4/thread-package

**Pull request merged**
